ext/quic/lib.deno_quic.d.ts

```diff
@@ -0,0 +1,160 @@
+   * ```
+   *
+   * Requires `allow-net` permission.
+   *
+   * @tags allow-net
+   * @category Network
+   */
+  export function connectQuic(options: ConnectQuicOptions): Promise<QuicConn>;
+
+  /** @category Network */
+  export interface QuicCloseInfo {
+    /** A number representing the error code for the error. */
+    closeCode: number;
+    /** A string representing the reason for closing the connection. */
+    reason: string;
+  }
+
+  /** Specialized listener that accepts QUIC connections.
+   *
+   * @category Network
+   */
+  export interface QuicListener extends AsyncIterable<QuicConn> {
+    /** Return the rid of the `QuicListener`. */
+    readonly rid: number;
+    /** Return the address of the `QuicListener`. */
+    readonly addr: NetAddr;
+
+    /** Waits for and resolves to the next connection to the `QuicListener`. */
+    accept(): Promise<QuicConn>;
+    /** Close closes the listener. Any pending accept promises will be rejected
+     * with errors. */
+    close(info: QuicCloseInfo): void;
+
+    [Symbol.asyncIterator](): AsyncIterableIterator<QuicConn>;
+  }
+
+  /** @category Network */
+  export interface QuicSendStreamOptions {
+    sendOrder?: number | undefined;
+  }
+
+  /** @category Network */
+  export interface QuicConn {
+    /** Close closes the listener. Any pending accept promises will be rejected
+     * with errors. */
+    close(info: QuicCloseInfo): void;
+    /** Opens and returns a bidirectional stream. */
+    createBidirectionalStream(
+      options?: QuicSendStreamOptions,
+    ): Promise<QuicBidirectionalStream>;
+    /** Opens and returns a unidirectional stream. */
+    createUnidirectionalStream(
+      options?: QuicSendStreamOptions,
+    ): Promise<QuicSendStream>;
+    /** Send a datagram. The provided data cannot be larger than
+     * `maxDatagramSize`. */
+    sendDatagram(data: Uint8Array): Promise<void>;
+    /** Receive a datagram. If no buffer is provider, one will be allocated.
+     * The zie of the provided buffer should be at least `maxDatagramSize`. */
+    readDatagram(buffer?: Uint8Array): Promise<Uint8Array>;
+
+    /** The resource ID of the connection. */
+    readonly rid: number;
+    /** Return the remote address for the connection. Clients may change
+     * addresses at will, e.g. when switching to a cellular internet connection.
+     */
+    readonly remoteAddr: NetAddr;
+    /** The negotiated ALPN protocol, if provided. */
+    readonly protocol: string | undefined;
+    /** Returns a promise that resolves when the connection is closed. */
+    readonly closed: Promise<QuicCloseInfo>;
+    /** A stream of bidirectional streams opened by the peer. */
+    readonly incomingBidirectionalStreams: ReadableStream<
+      QuicBidirectionalStream
+    >;
+    /** A stream of unidirectional streams opened by the peer. */
+    readonly incomingUnidirectionalStreams: ReadableStream<QuicReceiveStream>;
```

Comment:
I think it might be better to have a single incoming/outgoing stream type, which would allow for simpler async iteration over the listener.

Reply:
stream type is significant, for example when opening an http3 connection, each side needs to open a unidirectional "control" stream, and wait for the other side to open one.
i also don't think trying to write code that is generic over bidirectional and unidirectional streams would be very fun, since they're generally used for different purposes anyway.

Reply:
Looking at the WebTransport APIs, I see that they are using this same terminology, so sounds good.

Should this implement `datagrams` as well? https://developer.mozilla.org/en-US/docs/Web/API/WebTransport/datagrams

Reply:
i've been debating `datagrams`. there's a lot of complexity in that interface wrt how datagram packets are chunked through the streams, since datagrams don't match the normal read/write pattern of most resources, so i'm feeling like it might be better to leave that to the webtransport impl, and let this interface just have the raw send and read methods.